TargetComposition/SymBuilder/Readme.txt

```diff
@@ -273,6 +281,71 @@ Global data which is created with the "CreateGlobal" API has a set of available
 
         Delete            [Delete() - Deletes the global data]
 
+Functions which are are created with the "Create" API have a set of available properties as well
+
+    Function Objects
+    ----------------
+        Name              [The name of the symbol]
+        QualifiedName     [The qualified name of the symbol]
+        LocalVariables    [The list of local variables of the function]
+        Parameters        [The list of the parameters of the function]
+        ReturnType        [The return type of the function]
+
+The "Parameters" and "LocalVariables" properties on a function have methods to create new parameters or
+local variables within the function scope:
+
+    Parameters Object
+    -----------------
+        Add               [Add(name, parameterType) - Adds a new parameter of the given name and type]
+
+    LocalVariables Object
+    ---------------------
+        Add               [Add(name, localVariableType) - Adds a new local variable (non parameter) of the given name and type]
+
+Any local variable (whether a parameter or a not) has the following properties:
+
+    Parameter or LocalVariable Object
+    ---------------------------------
+        Name              [The name of the variable (parameter or local)]
+        QualifiedName     [The qualified name of the symbol]
+        Parent            [The parent of the symbol]
+        Type              [The type of the variable (parameter or local)]
+        LiveRanges        [The list of live ranges for the variable (parameter or local)]
+
+        Delete            [Delete() - Deletes the variable (parameter or local)]
+
+Parameters also have unique methods:
+
+    Parameter Object
+    ----------------
+        MoveBefore        [MoveBefore(pos) - Moves this parameter to a location in its parent function before the given parameter.  Note that 'pos' may either be a parameter object or it may be a zero based ordinal indicating its position within the list of parameters]
+
+The "LiveRanges" property of any variable is a list of the locations within the function where the variable is live
+and where that variable is located.  It also contains several APIs:
+
+    LiveRanges Object
+    -----------------
+        Add               [Add(rangeOffset, rangeSize, locDesc) - Adds a new live range for the variable within its containing function.  All offsets are function relative]
+
+Note that the live range offsets are function relative and must be within the bounds of the function.  The present
+extension also verifies that no two live ranges for the same variable overlap.  Location descriptions given

```

Comment:
Is that just a limitation currently? It seems like that could be useful, like in cases where a stack parameter gets temporarily enregistered.

Reply:
At the underlying interface level -- the debugger (or whatever client) creates a scope/scope frame for a given offset within the function and a singular location for a variable at that location is returned upwards.  How exactly that gets mapped is up to the implementation.

At this moment in time, the sample here chooses to not allow overlapping regions.  That's a present limitation of the sample.

That's not true of other providers like DWARF, etc... where there may indeed be overlapping regions...  some of which are control flow dependent and some of which are not.  At the end of the day, the plug-in chooses how to present the location for a variable at a given offset into the function.

I didn't want to overcomplicate the sample with that complexity at this point...

Reply:
Fair enough.